test/e2e/gateway_api_test.go

```diff
@@ -423,6 +428,286 @@ func testGatewayAPIRBAC(t *testing.T) {
+
+	if err := kclient.Update(context.TODO(), gateway); err != nil {
+		t.Fatalf("failed to update gateway %v: %v", gateway.Name, err)
+	}
+	t.Logf("Modified gateway's %s listener hostname to: %s", gateway.Name, "baz."+domain+".")
+
+	if err := assertExpectedDNSRecords(t, map[expectedDnsRecord]bool{
+		{dnsName: "baz." + domain + ".", gatewayName: "test-gateway-update"}: true,
+	}); err != nil {
+		t.Fatalf("DNSRecord expectations not met: %v", err)
+	}
+
+	// Fetch the latest version of the Gateway resource.
+	if err := kclient.Get(context.TODO(), types.NamespacedName{Namespace: operatorcontroller.DefaultOperandNamespace, Name: gateway.Name}, gateway); err != nil {
+		t.Fatalf("failed to get gateway resource: %v", err)
+	}
+
+	// Delete one of the listeners
+	var updatedListeners []gatewayapiv1.Listener
+	for _, l := range gateway.Spec.Listeners {
+		if l.Name == "http-listener2" {
+			continue
+		}
+		updatedListeners = append(updatedListeners, l)
+	}
+	gateway.Spec.Listeners = updatedListeners
+	if err := kclient.Update(context.TODO(), gateway); err != nil {
+		t.Fatalf("failed to update gateway with removed listener: %v", err)
+	}
+	t.Logf("Deleted listener %s from gateway %s", "http-listener2", gateway.Name)
+
+	t.Logf("Checking that the dnsRecord for %s gets removed.", "baz."+domain+".")
+
+	if err := assertExpectedDNSRecords(t, map[expectedDnsRecord]bool{
+		{dnsName: "baz." + domain + ".", gatewayName: "test-gateway-update"}: true,
+		{dnsName: "foo." + domain + ".", gatewayName: "test-gateway-update"}: false,
+		{dnsName: "bar." + domain + ".", gatewayName: "test-gateway-update"}: false,
+	}); err != nil {
+		t.Fatalf("DNSRecord expectations not met: %v", err)
+	}
+
+	// Delete the gateway
+	if err := kclient.Delete(context.TODO(), gateway); err != nil {
+		if errors.IsNotFound(err) {
+			return
+		}
+		t.Errorf("failed to delete gateway %q: %v", gateway.Name, err)
+	}
+
+	// Check the DNSRecord gets deleted for the remaining listener hostname
+	if err := assertExpectedDNSRecords(t, map[expectedDnsRecord]bool{
+		{dnsName: "baz." + domain + ".", gatewayName: "test-gateway-update"}: false,
+	}); err != nil {
+		t.Fatalf("DNSRecord expectations not met: %v", err)
+	}
+}
+
+func testGatewayAPIDNSListenerWithNoHostname(t *testing.T) {
+	gatewayClass, err := createGatewayClass(operatorcontroller.OpenShiftDefaultGatewayClassName, operatorcontroller.OpenShiftGatewayClassControllerName)
+	if err != nil {
+		t.Fatalf("failed to create gatewayclass: %v", err)
+	}
+
+	// Create a Gateway with the listener hostname "foo.[...]"
+	gateway, err := createGatewayWithListeners(gatewayClass, "test-nohost-gateway", operatorcontroller.DefaultOperandNamespace, []struct {
+		Name     string
+		Hostname *string
+	}{
+		{Name: "http-listener-no-host", Hostname: nil},
+	})
+
+	t.Cleanup(func() {
+		gateway := gatewayapiv1.Gateway{ObjectMeta: metav1.ObjectMeta{Name: "test-nohost-gateway", Namespace: operatorcontroller.DefaultOperandNamespace}}
```

Comment:
nit: you shouldn't need to recreate `gateway`, the original should still be around.